### ISSUE TITLE
Metabase slow on 0.39.2+ on Kubernetes

### ISSUE THREAD

**Issue by damovsky** [metabase/metabase#16669]
**Describe the bug**
Metabase is often very slow and users reports it's down. This problem started after migration from 0.39.0.1 to 0.39.3. We tried to upgrade to 0.39.4 and issue still persist. Our monitoring in datadog confirmed the issue there.

Our synthetic tests in datadog does following:
- HTTP GET every 5 minutes to URL <metabase>/auth/login?redirect=%2F
-  we can see the time increased from hundreds of ms to seconds (the max measured time was 11 seconds just to get reponse from auth page of metabase)
- we measure first byte response time

this is graph from datadog for past month...we did upgrade on 11th of Jun
![image](https://user-images.githubusercontent.com/8916256/122550932-9b75d680-d034-11eb-992b-1e5eba955813.png)

We do not see any spike in CPU load nor in RAM usage during the time of slow metabase response

**Logs**
will attach logs once I have something...

**To Reproduce**
cannot reproduce

**Expected behavior**
response time is within hundreds of ms

**Screenshots**
already provided

**Information about your Metabase Installation:**

- We run metabase in k8s cluster in AWS, with using HELM chart 
- As a metabase DB we use AWS RDS Postgresql. 
- We use Google auth for user authentication.

```{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.101 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.11+9",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.11",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.11+9",
    "os.name": "Linux",
    "os.version": "5.4.95-42.163.amzn2.x86_64",
    "user.language": "en",
    "user.timezone": "UTC"
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "snowflake",
      "mysql"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "12.5"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.18"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2021-06-15",
      "tag": "v0.39.4",
      "branch": "release-x.39.x",
      "hash": "f538050"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**
It's annoying - we have to restart metabase very often as a quick fix

**Additional context**
n/a

**Comment by flamber** [comment]
Hi @damovsky 
Can you provide your k8s configuration? Which ingress do you use?
The Metabase logs would be very helpful in determining which resources are available.
Can you try 0.39.1 - I have seen similar report of slowness, which only seems to occur on 0.39.2+, but we haven't been able to track down the root cause of this, since there hasn't been enough information on how to reproduce similar environment.

**Comment by damovsky** [comment]
hi @flamber,
is it safe to downgrade from 0.39.4 to 0.39.1 ? there are some DB migrations as far as I know...

**Comment by flamber** [comment]
@damovsky There's aren't any migrations between those two versions, so it shouldn't be a problem - but as always, backup first.

**Comment by damovsky** [comment]
just having the perf issue right now....here are logs and graph with response time

[metabase-logs.txt](https://github.com/metabase/metabase/files/6677527/metabase-logs.txt)

![image](https://user-images.githubusercontent.com/8916256/122574410-ea7d3500-d04f-11eb-8053-aac1c12eb7e7.png)


will do downgrade next week.

**Comment by flamber** [comment]
@damovsky 
Can you also check the Metabase application database (PostgreSQL 12.5) for high load (many queries, high CPU, etc)?
Which other environment variables, besides the [app db pool](https://www.metabase.com/docs/latest/operations-guide/environment-variables.html#mb_application_db_max_connection_pool_size) have you changed?

**Comment by damovsky** [comment]
just downgraded to 0.39.4. Will be monitoring it today...

I don't see any high cpu load nor query increase during the time of latency spike on friday

![image](https://user-images.githubusercontent.com/8916256/122728789-c2244f00-d278-11eb-85fb-d6110ec2db5f.png)


Regarding the env variables - this is the only change we did as far as I know
`  javaOpts: "-Dnetworkaddress.cache.ttl=1 -Dc3p0.minPoolSize=1 -Dc3p0.maxPoolSize=20 -Dc3p0.maxIdleTimeExcessConnections=600 -Dc3p0.maxConnectionAge=900"
`

**Comment by flamber** [comment]
@damovsky Very interesting. I guess you meant 0.39.1. Please report back later with a status if you see spikes.

`minPoolSize` is already defaulted to 1 ([ref1](https://github.com/metabase/metabase/blob/ce939219f8a3e28ecb00deb232154780a1dcfcef/resources/c3p0.properties#L2), [ref2](https://github.com/metabase/metabase/blob/478ec730cce3f17cae9646c98422a409d47c9f02/src/metabase/driver/sql_jdbc/connection.clj#L52)).
`maxPoolSize` is defaulted to 15 ([ref](https://github.com/metabase/metabase/blob/478ec730cce3f17cae9646c98422a409d47c9f02/src/metabase/driver/sql_jdbc/connection.clj#L54-L55)), but I don't see that you even reach this amount from your log, so unsure if it's even needed, but perhaps your log didn't show your instance under full load.
`maxIdleTimeExcessConnections` is defaulted to 300 since 0.38.4 ([ref](https://github.com/metabase/metabase/blob/478ec730cce3f17cae9646c98422a409d47c9f02/src/metabase/driver/sql_jdbc/connection.clj#L79)) - used to be 900 before
`maxConnectionAge` is not set in Metabase, which means it defaults to [c3p0](https://www.mchange.com/projects/c3p0/#maxConnectionAge)

It would be extremely helpful if you could provide your k8s config, since if we cannot reproduce, then it's a lot harder to fix something blindly.

**Comment by damovsky** [comment]
yes, correct...downgraded to 0.39.1 and it works much better than with 0.39.4!

<img width="574" alt="Screenshot 2021-06-21 at 16 32 35" src="https://user-images.githubusercontent.com/8916256/122779701-6674b880-d2ae-11eb-8dec-906e46e7ff8b.png">

what kind of k8s config would you like? I can't give you half of our infra repo code 

[ingress.txt](https://github.com/metabase/metabase/files/6692735/ingress.txt)

**Comment by flamber** [comment]
@damovsky Thank you for testing! And great that we're now getting closer to finding the culprit.
I'm of course only interested in any configs related to Metabase, so we can try to reproduce similar setup. I don't want you to post anything you don't want, but if you prefer, you can email to my username `@metabase.com`

**Comment by lindsay-stevens** [comment]
@flamber I also have poor performance on Metabase v0.39.2. Using CentOS 7 with Postgres 10 back-end and warehouse. The instance has 20+ Metabase databases, which are synced hourly or so. Usually that is fine, but with v0.39.2 the instance slows to a crawl and the Metabase UI is unusable. For example, 5 seconds for /api/current (loaded every page), or 30 seconds to get /api/databases (the Admin -> Databases page). The back-end databases are functioning normally; watching incoming Metabase queries with `pg_stat_activity` then running them manually, the queries execute in sub-second times. For example `select * from metabase_database` in 9ms.

Given the reports of better performance in v0.39.1, I had a look at the changes between versions. It seems that in #15916 there was a single thread threadpool added for table metadata updates. In other words, a bottleneck for any sync activity across the instance. Maybe the wider performance impact across the app is due to other parts of the code waiting on sync activity to complete. So then perhaps the reproduce this issue involves running multiple concurrent database syncs, and then trying to use the UI / call the API.

**Comment by dpsutton** [comment]
>  It seems that in #15916 there was a single thread threadpool added for table metadata updates. In other words, a bottleneck for any sync activity across the instance.

Hi @lindsay-stevens . This threadpool was added for resyncing individual tables when they are unhidden in the UI. The only usage of that threadpool is in the endpoints PUT `api/table/:id` or `api/table` and then only when moving one or many tables from `visibility_type` "hidden" to the normal (nil) value. (only call site here: https://github.com/metabase/metabase/blob/master/src/metabase/api/table.clj#L86)

The periodic scheduled syncing of databases remains untouched and occurs in https://github.com/metabase/metabase/blob/master/src/metabase/task/sync_databases.clj#L58-L75. These jobs are scheduled by Quartz on a ten thread threadpool.

```clojure
(.. (org.quartz.impl.StdSchedulerFactory/getDefaultScheduler) getMetaData getThreadPoolSize)
10
```

I'm gonna research this tomorrow and I want to track down what is going on but I don't believe this change is related.

**Comment by lindsay-stevens** [comment]
Hi @dpsutton thanks for looking in to this. Previous Metabase version was v0.35.4 if that helps. Also, I was testing this further and saw that for every table being un-hidden, a new database connection is opened / closed, presumably to service the `submit-task` call to `sync/sync-table!`. Could this be contributing to the performance cost in Metabase?

**Comment by dpsutton** [comment]
@lindsay-stevens are you hiding and unhiding lots of tables? My gut reaction is this is not a frequent action so it is not a significant contributor to the reports of general slowness on k8s. We use a connection pool (c3p0) so we should be intelligently reusing connections up to a max limit. As to this being a factor in the slowdown, I doubt it but open to evidence of it.

As for the slowdown, do you think it is possible for you to get some stack dumps from the instances in k8s? Attaching visualvm or some other profiler/logger, or if you can get access to a shell on the pod, using `jstack` to get a stackdump would be super helpful.

**Comment by lindsay-stevens** [comment]
@dpsutton I was tailing the postgres logs with `log_connections = on` and the hide/un-hide action was associated with a new connection each time. Would it be feasible to separate the limit by database? So an instance with 20 databases isn't limited to 1 table show/hide at a time, but can do 20.

Anyway, I did some further testing and I think this might instead be related to the Metabase database sync step `sync-fields`. Relevant log output below. It seems that somewhere between v0.35.4 and v0.39.2, that `sync-fields` step takes twice as long with the same data warehouses. I haven't copied the logs but results were similar between v0.39.1 and v0.39.2. There were a lot of changes between v0.35.4 and v0.39.2 so I'm not sure where the issue came from.

In my situation, there is an external application managing the Metabase configuration via the Metabase API. The application does things like setting up new databases, maintaining the data reference, creating template / starter questions, etc. This work occurs every hour or less, and as part of that process a database sync is requested. Instances typically have 2 to 8 or sometimes ~50 databases. Each database typically has 50 to 100 tables, and total 2k or more fields across those tables. So a 2x slow-down in `sync-fields` would snowball into a big resource usage impact. During sync, Metabase uses ~50% of a CPU with spikes between 100% and 150%.

Are these results reproducible on your end?

Note that each sync is a manually triggered sync (`api/database/:id/sync_schema`), not the one that runs automatically straight after adding a database.

<details>
  <summary>v0.35.4, fresh Metabase db, custom test data warehouse (Postgres). Total sync time 2s, sync-fields step 1s.</summary>

```
2021-07-08 19:54:50 | INFO sync.util :: STARTING: Sync metadata for my-custom-postgres-driver Database 3 'my__test_database'
2021-07-08 19:54:50 | INFO sync.util :: STARTING: step 'sync-timezone' for my-custom-postgres-driver Database 3 'my__test_database'
2021-07-08 19:54:50 | INFO sync.util :: FINISHED: step 'sync-timezone' for my-custom-postgres-driver Database 3 'my__test_database' (4.0 ms)
2021-07-08 19:54:50 | INFO sync.util :: STARTING: step 'sync-tables' for my-custom-postgres-driver Database 3 'my__test_database'
2021-07-08 19:54:50 | INFO sync.util :: FINISHED: step 'sync-tables' for my-custom-postgres-driver Database 3 'my__test_database' (11.4 ms)
2021-07-08 19:54:50 | INFO sync.util :: STARTING: step 'sync-fields' for my-custom-postgres-driver Database 3 'my__test_database'
2021-07-08 19:54:51 | INFO sync.util :: FINISHED: step 'sync-fields' for my-custom-postgres-driver Database 3 'my__test_database' (966.2 ms)
2021-07-08 19:54:51 | INFO sync.util :: STARTING: step 'sync-fks' for my-custom-postgres-driver Database 3 'my__test_database'
2021-07-08 19:54:52 | INFO sync.util :: FINISHED: step 'sync-fks' for my-custom-postgres-driver Database 3 'my__test_database' (1.1 s)
2021-07-08 19:54:52 | INFO sync.util :: STARTING: step 'sync-metabase-metadata' for my-custom-postgres-driver Database 3 'my__test_database'
2021-07-08 19:54:52 | INFO sync.util :: FINISHED: step 'sync-metabase-metadata' for my-custom-postgres-driver Database 3 'my__test_database' (9.7 ms)
2021-07-08 19:54:52 | INFO sync.util :: FINISHED: Sync metadata for my-custom-postgres-driver Database 3 'my__test_database' (2.1 s)
```
</details>

<details>
  <summary>v0.39.2, fresh Metabase db, upgraded from v0.35.4, copy of same custom data warehouse (Postgres). Total sync time 5s, sync-fields step 3.8s.</summary>

```
2021-07-08 10:07:34,575 | INFO | metabase.sync.util | STARTING: Sync metadata for my-custom-postgres-driver Database 4 'my__test_database'
2021-07-08 10:07:34,576 | INFO | metabase.sync.util | STARTING: step 'sync-timezone' for my-custom-postgres-driver Database 4 'my__test_database'
2021-07-08 10:07:34,579 | INFO | metabase.sync.analyze | classify-fields Analyzed [*******************************···················]    62% Table 196 'my_schema.table_1'
2021-07-08 10:07:34,583 | INFO | metabase.sync.util | FINISHED: step 'sync-timezone' for my-custom-postgres-driver Database 4 'my__test_database' (7.1 ms)
2021-07-08 10:07:34,583 | INFO | metabase.sync.util | STARTING: step 'sync-tables' for my-custom-postgres-driver Database 4 'my__test_database'
2021-07-08 10:07:34,602 | INFO | metabase.sync.util | FINISHED: step 'sync-tables' for my-custom-postgres-driver Database 4 'my__test_database' (18.6 ms)
2021-07-08 10:07:34,603 | INFO | metabase.sync.util | STARTING: step 'sync-fields' for my-custom-postgres-driver Database 4 'my__test_database'
2021-07-08 10:07:35,068 | INFO | metabase.sync.util | FINISHED: step 'classify-fields' for my-custom-postgres-driver Database 4 'my__test_database' (6.4 s)
2021-07-08 10:07:35,068 | INFO | metabase.sync.util | STARTING: step 'classify-tables' for my-custom-postgres-driver Database 4 'my__test_database'
2021-07-08 10:07:35,174 | INFO | metabase.sync.analyze | classify-tables Analyzed [************************************··············]    73% Table 144 'my_schema.table_2'
2021-07-08 10:07:35,345 | INFO | metabase.sync.analyze | classify-tables Analyzed [*****************************************·········]    83% Table 164 'my_schema.table_3'
2021-07-08 10:07:35,490 | INFO | metabase.sync.analyze | classify-tables Analyzed [**********************************************····]    93% Table 184 'my_schema.table_4'
2021-07-08 10:07:35,581 | INFO | metabase.sync.util | FINISHED: step 'classify-tables' for my-custom-postgres-driver Database 4 'my__test_database' (512.2 ms)
2021-07-08 10:07:35,635 | INFO | metabase.sync.util | FINISHED: Analyze data for my-custom-postgres-driver Database 4 'my__test_database' (29.4 s)
2021-07-08 10:07:38,370 | INFO | metabase.sync.util | FINISHED: step 'sync-fields' for my-custom-postgres-driver Database 4 'my__test_database' (3.8 s)
2021-07-08 10:07:38,370 | INFO | metabase.sync.util | STARTING: step 'sync-fks' for my-custom-postgres-driver Database 4 'my__test_database'
2021-07-08 10:07:39,535 | INFO | metabase.sync.util | FINISHED: step 'sync-fks' for my-custom-postgres-driver Database 4 'my__test_database' (1.2 s)
2021-07-08 10:07:39,536 | INFO | metabase.sync.util | STARTING: step 'sync-metabase-metadata' for my-custom-postgres-driver Database 4 'my__test_database'
2021-07-08 10:07:39,548 | INFO | metabase.sync.util | FINISHED: step 'sync-metabase-metadata' for my-custom-postgres-driver Database 4 'my__test_database' (12.4 ms)
2021-07-08 10:07:39,588 | INFO | metabase.sync.util | FINISHED: Sync metadata for my-custom-postgres-driver Database 4 'my__test_database' (5.0 s)
```
</details>

<details>
  <summary>v0.35.4, fresh Metabase db, sample database (H2). Total sync time 315ms, sync-fields 55ms</summary>

```
2021-07-08 20:29:38 | INFO sync.util :: STARTING: Sync metadata for h2 Database 1 'Sample Dataset'
2021-07-08 20:29:38 | INFO sync.util :: STARTING: step 'sync-timezone' for h2 Database 1 'Sample Dataset'
2021-07-08 20:29:38 | INFO sync.util :: FINISHED: step 'sync-timezone' for h2 Database 1 'Sample Dataset' (6.1 ms)
2021-07-08 20:29:38 | INFO sync.util :: STARTING: step 'sync-tables' for h2 Database 1 'Sample Dataset'
2021-07-08 20:29:38 | INFO sync-metadata.tables :: Updating description for tables: (Table  'PUBLIC.PEOPLE' Table  'PUBLIC.REVIEWS' Table  'PUBLIC.ORDERS' Table  'PUBLIC.PRODUCTS')
2021-07-08 20:29:38 | INFO sync.util :: FINISHED: step 'sync-tables' for h2 Database 1 'Sample Dataset' (9.8 ms)
2021-07-08 20:29:38 | INFO sync.util :: STARTING: step 'sync-fields' for h2 Database 1 'Sample Dataset'
2021-07-08 20:29:38 | INFO sync.util :: FINISHED: step 'sync-fields' for h2 Database 1 'Sample Dataset' (55.2 ms)
2021-07-08 20:29:38 | INFO sync.util :: STARTING: step 'sync-fks' for h2 Database 1 'Sample Dataset'
2021-07-08 20:29:38 | INFO sync.util :: FINISHED: step 'sync-fks' for h2 Database 1 'Sample Dataset' (9.3 ms)
2021-07-08 20:29:38 | INFO sync.util :: STARTING: step 'sync-metabase-metadata' for h2 Database 1 'Sample Dataset'
2021-07-08 20:29:38 | INFO sync.util :: FINISHED: step 'sync-metabase-metadata' for h2 Database 1 'Sample Dataset' (206.8 ms)
2021-07-08 20:29:38 | INFO sync.util :: FINISHED: Sync metadata for h2 Database 1 'Sample Dataset' (315.1 ms)
```
</details>

<details>
  <summary>v0.39.2, fresh Metabase db, upgraded from v0.35.4, sample database (H2). Total time 415ms, sync-fields step 108ms.</summary>

```
2021-07-08 10:21:47,052 | INFO | metabase.sync.util | STARTING: Sync metadata for h2 Database 1 'Sample Dataset'
2021-07-08 10:21:47,053 | INFO | metabase.sync.util | STARTING: step 'sync-timezone' for h2 Database 1 'Sample Dataset'
2021-07-08 10:21:47,060 | INFO | metabase.sync.util | FINISHED: step 'sync-timezone' for h2 Database 1 'Sample Dataset' (7.5 ms)
2021-07-08 10:21:47,061 | INFO | metabase.sync.util | STARTING: step 'sync-tables' for h2 Database 1 'Sample Dataset'
2021-07-08 10:21:47,076 | INFO | metabase.sync.sync-metadata.tables | Updating description for tables: (Table  'PUBLIC.PEOPLE' Table  'PUBLIC.REVIEWS' Table  'PUBLIC.ORDERS' Table  'PUBLIC.PRODUCTS')
2021-07-08 10:21:47,077 | INFO | metabase.sync.util | FINISHED: step 'sync-tables' for h2 Database 1 'Sample Dataset' (15.7 ms)
2021-07-08 10:21:47,077 | INFO | metabase.sync.util | STARTING: step 'sync-fields' for h2 Database 1 'Sample Dataset'
2021-07-08 10:21:47,185 | INFO | metabase.sync.util | FINISHED: step 'sync-fields' for h2 Database 1 'Sample Dataset' (108.4 ms)
2021-07-08 10:21:47,186 | INFO | metabase.sync.util | STARTING: step 'sync-fks' for h2 Database 1 'Sample Dataset'
2021-07-08 10:21:47,199 | INFO | metabase.sync.util | FINISHED: step 'sync-fks' for h2 Database 1 'Sample Dataset' (13.0 ms)
2021-07-08 10:21:47,200 | INFO | metabase.sync.util | STARTING: step 'sync-metabase-metadata' for h2 Database 1 'Sample Dataset'
2021-07-08 10:21:47,479 | INFO | metabase.sync.util | FINISHED: step 'sync-metabase-metadata' for h2 Database 1 'Sample Dataset' (279.5 ms)
2021-07-08 10:21:47,504 | INFO | metabase.sync.util | FINISHED: Sync metadata for h2 Database 1 'Sample Dataset' (451.4 ms)
```
</details>

**Comment by Oreilles** [comment]
Facing similar issue here: questions that would take 5 seconds to run in 0.39.1 could take up to a full minute in 0.39.4.
Downgrading solved the problem.

**Comment by lindsay-stevens** [comment]
Might be explained by logging performance finding in https://github.com/metabase/metabase/issues/16380#issuecomment-878731691 with PR #17002

**Comment by flamber** [comment]
Everyone, please try 0.39.5 or 0.40.1 released a few hours ago, which has the PR that @lindsay-stevens referenced.
As always, remember to backup before upgrading, so you can revert in case of problems.
https://github.com/metabase/metabase/releases

**Comment by elsbrock** [comment]
0.39.5 solved the problem for us (we upgraded from 0.39.3)

**Comment by flamber** [comment]
@elsbrock Thank you.
@lindsay-stevens Did it also resolve what you were seeing? If not, then I think you might be seeing a different sync specific issue.

I'm going to close this as fixed in 0.39.5 and 0.40.1+, since I have not seen any other similar reports the past couple of weeks.

If you are (or think you are) still experiencing this issue, then please provide enough information about your setup and monitoring, so we have a chance to investigate (a comment like "I'm still seeing this" provides no useful help).

**Comment by lindsay-stevens** [comment]
@flamber this seems to be improved / resolved in v0.40.2. I have some Metabase client tests that were running in 10-15 mins on v0.39.2 which are back down to ~5mins on v0.40.2. The issue might crop up later in performance test scenarios etc but so far so good.
